src/main/java/org/jitsi/jicofo/AbstractChannelAllocator.java

```diff
@@ -0,0 +1,394 @@
+     * The class logger which can be used to override logging level inherited
+     * from {@link JitsiMeetConference}.
+     */
+    private final static Logger classLogger
+        = Logger.getLogger(AbstractChannelAllocator.class);
+
+    /**
+     * The logger for this instance. Uses the logging level either of the
+     * {@link #classLogger} or {@link JitsiMeetConference#getLogger()}
+     * whichever is higher.
+     */
+    private final Logger logger;
+
+    /**
+     * The {@link JitsiMeetConferenceImpl} into which a participant will be
+     * invited.
+     */
+    protected final JitsiMeetConferenceImpl meetConference;
+
+    /**
+     * The {@link JitsiMeetConferenceImpl.BridgeSession} on which
+     * to allocate channels for the participant.
+     */
+    protected final JitsiMeetConferenceImpl.BridgeSession bridgeSession;
+
+    /**
+     * The participant that is to be invited by this instance to the conference.
+     */
+    protected final Participant participant;
+
+    /**
+     * A flag which indicates whether channel allocation is canceled. Raising
+     * this makes the allocation thread discontinue the allocation process and
+     * return.
+     */
+    protected volatile boolean canceled = false;
+
+    /**
+     * First argument stands for "start audio muted" and the second one for
+     * "start video muted". The information is included as a custom extension in
+     * 'session-initiate' sent to the user.
+     */
+    protected final boolean[] startMuted;
+
+    /**
+     * Indicates whether or not this thread will be doing a "re-invite". It
+     * means that we're going to replace previous conference which has failed.
+     * Channels are allocated on new JVB and peer is re-invited with
+     * 'transport-replace' Jingle action as opposed to 'session-initiate' in
+     * regular invite.
+     */
+    protected final boolean reInvite;
+
+    /**
+     * Initializes a new {@link AbstractChannelAllocator} instance which is to
+     * invite a specific {@link Participant} into a specific
+     * {@link JitsiMeetConferenceImpl} (using a specific jitsi-videobridge
+     * instance specified by a
+     * {@link org.jitsi.jicofo.JitsiMeetConferenceImpl.BridgeSession}).
+     *
+     * @param meetConference the {@link JitsiMeetConferenceImpl} into which to
+     * invite {@code participant}.
+     * @param bridgeSession the
+     * {@link org.jitsi.jicofo.JitsiMeetConferenceImpl.BridgeSession} which
+     * identifies the jitsi-videobridge instance on which to allocate channels.
+     * @param participant the participant to be invited.
+     * @param startMuted an array which must have the size of 2 where the first
+     * value stands for "start audio muted" and the second one for "video
+     * muted". This is to be included in client's offer.
+     * @param reInvite whether to send an initial offer (session-initiate) or
+     * a an updated offer (transport-replace).
+     */
+    protected AbstractChannelAllocator(
```

Comment:
They could be separated. I would like to avoid doing it because it's a complex piece and it's tricky to modify without breaking it. And I actually like that the whole task (which we run in a separate thread) is all contained in this class (and its subclasses).
If you feel strongly about it I'll try to separate it somehow.